crates/admin/src/storage_query/convert.rs

```diff
@@ -0,0 +1,281 @@
+            TimestampMillisecondType,
+        },
+        error::ArrowError,
+    },
+    error::DataFusionError,
+    execution::{RecordBatchStream, SendableRecordBatchStream},
+};
+use futures::{Stream, StreamExt};
+
+pub(super) const V1_CONVERTER: JoinConverter<
+    JoinConverter<LargeConverter, FullCountConverter>,
+    TimestampConverter,
+> = JoinConverter::new(
+    JoinConverter::new(LargeConverter, FullCountConverter),
+    TimestampConverter,
+);
+
+pub(super) struct ConvertRecordBatchStream<C> {
+    converter: C,
+    inner: SendableRecordBatchStream,
+    converted_schema: SchemaRef,
+    done: bool,
+}
+
+impl<C: Converter> ConvertRecordBatchStream<C> {
+    pub(super) fn new(converter: C, inner: SendableRecordBatchStream) -> Self {
+        let converted_schema = converter.convert_schema(inner.schema());
+
+        Self {
+            converter,
+            inner,
+            converted_schema,
+            done: false,
+        }
+    }
+}
+
+impl<C: Converter> RecordBatchStream for ConvertRecordBatchStream<C> {
+    fn schema(&self) -> SchemaRef {
+        self.converted_schema.clone()
+    }
+}
+
+impl<C: Converter> Stream for ConvertRecordBatchStream<C> {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.done {
+            return Poll::Ready(None);
+        }
+
+        let record_batch = ready!(self.inner.poll_next_unpin(cx));
+
+        Poll::Ready(match record_batch {
+            Some(record_batch) => Some(match record_batch {
+                Ok(record_batch) => Ok(self
+                    .converter
+                    .convert_record_batch(&self.converted_schema, record_batch)?),
+                Err(err) => {
+                    self.done = true;
+                    Err(err)
+                }
+            }),
+            None => None,
+        })
+    }
+}
+
+pub(super) trait Converter: Unpin {
+    fn convert_schema(&self, schema: SchemaRef) -> SchemaRef {
+        let fields = Vec::from_iter(schema.fields().iter().cloned());
+        let fields = self.convert_fields(fields);
+        SchemaRef::new(Schema::new_with_metadata(fields, schema.metadata().clone()))
+    }
+
+    fn convert_record_batch(
+        &self,
+        converted_schema: &SchemaRef,
+        record_batch: RecordBatch,
+    ) -> Result<RecordBatch, ArrowError> {
+        let columns = Vec::from_iter(record_batch.columns().iter().cloned());
+        let columns = self.convert_columns(converted_schema, columns)?;
+
+        RecordBatch::try_new(converted_schema.clone(), columns)
+    }
+
+    fn convert_columns(
+        &self,
+        converted_schema: &SchemaRef,
+        columns: Vec<ArrayRef>,
+    ) -> Result<Vec<ArrayRef>, ArrowError>;
+
+    fn convert_fields(&self, fields: Vec<FieldRef>) -> Vec<FieldRef>;
+}
+
+pub(super) struct JoinConverter<Before, After> {
+    before: Before,
+    after: After,
+}
```

Comment:
I tripped a bit about the names before and after here. I guess it encodes a chain of conversions, right? Maybe call it `first` and `second`?

Reply:
sure